Wikipedia/Code/WMFDatabaseHouseKeeper.swift

```diff
@@ -0,0 +1,97 @@
+        let thirtyDaysAgoMidnightUTC = utcCalendar.date(byAdding: .day, value: -30, to: midnightTodayUTC, wrappingComponents: true)
+        
+        let allContentGroupFetchRequest = WMFContentGroup.fetchRequest()
+        allContentGroupFetchRequest.propertiesToFetch = ["key", "midnightUTCDate", "content", "articleURLString"]
+        
+        let allContentGroups = try moc.fetch(allContentGroupFetchRequest)
+        var referencedArticleKeys = Set<String>(minimumCapacity: allContentGroups.count * 5 + 1)
+        
+        for group in allContentGroups {
+            if group.midnightUTCDate?.compare(thirtyDaysAgoMidnightUTC!) == .orderedAscending {
+                moc.delete(group)
+                continue
+            }
+            
+            if let articleURLDatabaseKey = (group.articleURL as NSURL?)?.wmf_articleDatabaseKey {
+                referencedArticleKeys.insert(articleURLDatabaseKey)
+            }
+            
+            guard let content = group.content as Array! else {
+                assertionFailure("Unknown Content Type")
+                continue
+            }
+            
+            for obj in content {
+                
+                switch (group.contentType, obj) {
+                    
+                case (.URL, let url as NSURL):
+                    guard let key = url.wmf_articleDatabaseKey else {
+                        continue
+                    }
+                    referencedArticleKeys.insert(key)
+                    
+                case (.topReadPreview, let preview as WMFFeedTopReadArticlePreview):
+                    guard let key = (preview.articleURL as NSURL).wmf_articleDatabaseKey else {
+                        continue
+                    }
+                    referencedArticleKeys.insert(key)
+                    
+                case (.story, let story as WMFFeedNewsStory):
+                    if let articlePreviews = story.articlePreviews {
+                        for preview in articlePreviews {
+                            guard let key = (preview.articleURL as NSURL).wmf_articleDatabaseKey else {
+                                continue
+                            }
+                            referencedArticleKeys.insert(key)
+                        }
+                    }
+                    
+                case (.URL, _),
+                     (.topReadPreview, _),
+                     (.story, _),
+                     (.image, _),
+                     (.announcement, _):
+                    break
+                    
+                default:
+                    assertionFailure("Unknown Content Type")
+                }
+            }
+        }
+        
+        let articlesToDeleteFetchRequest = WMFArticle.fetchRequest()
+        var articlesToDeletePredicate = NSPredicate(format: "viewedDate == NULL && savedDate == NULL && placesSortOrder == NULL && isExcludedFromFeed == %@", NSNumber(booleanLiteral: true))
```

Comment:
A `viewedDate` of null indicates that the article was never viewed, and is only a preview. `savedDate` of null indicates the article isn't saved. Saved and history are populated by looking for articles with viewed dates and saved dates. A non-null `placesSortOrder` indicates that the article is currently visible on the places map. `isExcludedFromFeed` here is incorrect - it should check for items not excluded from the feed. Items with `isExcludedFromFeed == YES` need to stay in the database so that they will continue to be excluded from the feed. Previous code: ```NSPredicate *articlesToDeletePredicate = [NSPredicate predicateWithFormat:@"viewedDate == NULL && savedDate == NULL && placesSortOrder == NULL && isExcludedFromFeed == %@", @(NO)];```